### PR TITLE
add cache option for additive and linear trainer

### PR DESCRIPTION
**Pull request by deerzq**
- add cache option to allow faster training on small dataset
- remove the isRanking option in additive model trainer since it's not used

to: @jq @spencerde

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 1]
```diff
@@ -299,17 +311,6 @@ object AdditiveModelTrainer {
     return loss
   }
 
-  private def transformExamples(input: RDD[Example],
-                        config: Config,
-                        key: String,
-                        params: AdditiveTrainerParams): RDD[Example] = {
-    if (params.isRanking) {
-      LinearRankerUtils.transformExamples(input, config, key)
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: so we don't do isRanking, what's the difference between transformExamples and makePointwiseFloat, both do combineContextAndItems, but transformExamples unsetContext, what's the role of Context play here?

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 2]
```diff
@@ -52,7 +52,13 @@ object AdditiveModelTrainer {
     val trainConfig = config.getConfig(key)
     val iterations : Int = trainConfig.getInt("iterations")
     val params = loadTrainingParameters(trainConfig)
-    val transformed = transformExamples(input, config, key, params)
+    val pointwise = LinearRankerUtils.makePointwiseFloat(input, config, key)
+
+    val transformed = params.cache match {
+      case "memory" => pointwise.cache()
```
Comment (jq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: nice! this should boost the performance!

**Comment by jq** [comment]
Just a knowledge question LGTM

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala (hunk 1 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: when `makePointwiseFloat` is used one Example contains only one example, but when `transformExamples` is used one Example can contain multiple examples and it is used for compute ranking loss, if you look at the `getTrainingExamples` in PricingPipeline, you can see how examples are grouped. As for context, I think for both cases we are not using it.

**Pull request merged**